Sync the product attributes lookup table on product creation/change/deletion

Konamiman opened the pull request:
### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Syncrhonizes the product attributes lookup table whenever a product or variation is created, modified or deleted. Syncrhonization happens whenever the lookup table exists, even if the usage of the table isn't active. The filtering is showcased in #29190, the initial table creation and filling is implemented in #29778, and the actual usage of the table for filtering is implemented in #29896.

Entries in the lookup table are created when:

* A new product is created
* A new variation is added to a product
* A product or variation that had been trashed is untrashed
* The product visibility changes to "search" or "hidden"
* The attributes for a product change, this includes:
  * New attributes are added or deleted for the product
  * Terms are added or removed from the attribute definition for the product
  * The attribute changes between "used for variations" and not
* The attributes for a variation change

Entries in the lookup table are deleted when:

* A product  or variation is trashed or permanently deleted
* The product visibility changes to "visible" or "catalog"

Entries in the lookup table are updated only when the stock status of a product or variation changes.

#### Scheduled vs direct updates

By default changes to the lookup table are not performed directly when the product changes, but a scheduled action is used instead. The name of the hook is `woocommerce_run_product_attribute_lookup_update_callback` (the `DataRegenerator` class used a hook with the same name, which has been renamed to `woocommerce_run_product_attribute_lookup_regeneration_callback`).

WooCommerce is prone to save stuff multiple times in a row, therefore a check is done before scheduling the action to prevent duplicates (existing pending action for same product and type of action, create/update/delete).

A new option is introduced, `woocommerce_attribute_lookup__direct_updates`. When set to `yes` updates are done directly when the change happens, skipping action scheduler; this may be useful for testing.

#### Per product regeneration

The "Regenerate the product attributes lookup table" tool in the Status - Tools page has been extended to include a product selector. If a product is selected then the "Regenerate" button will regenerate the data only for the product selected (directly, without scheduled action); if the selector is left blank then the table is fully regenerated using scheduled actions, as it was doing before.

![image](https://user-images.githubusercontent.com/937723/125607277-b5a9d837-0779-423c-b276-0e1429518f2a.png)

The single product regeneration may be useful if for some reason the lookup data ends up being incorrect for some reason, and ideally it would never be needed.

#### Changes in settings UI

When the feature is enabled (via `wc_get_container()->get(\Automattic\WooCommerce\Internal\ProductAttributesLookup\LookupDataStore::class)->show_feature();`) and the lookup table exists a new "Advanced" section is added to the product settings:

![image](https://user-images.githubusercontent.com/937723/124109557-90c13580-da67-11eb-8d17-50fd348455eb.png)

If a data regeneration is in progress then the section is still displayed but with no actionable options:

![image](https://user-images.githubusercontent.com/937723/124109913-f0b7dc00-da67-11eb-93f9-f3947a98d044.png)

Closes #29609.

### How to test the changes in this Pull Request:

1. Create and fill the lookup table according to the instructions of #29190.
2. Create/delete/modify products and variations according to the cases listed above, and verify that the lookup table updates accordingly.
3. Delete the lookup table (from the tools page) and verify that everything continues working.
4. To test the single product regeneration you can manually delete the lookup data for the product before using the regeneration tool and then check that the data was regenerated.

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Add - Product attributes lookup table synchronization when the table exists.

james-allan committed:
Update the rate when recalculating and updating order tax items

claudiosanches committed:
Merge pull request #27985 from woocommerce/set_rate_id_on_recalculate

Update the tax rate meta when recalculating and updating order tax items

Konamiman committed:
Refactor save method in WC_Product and WC_Product_Variable

The 'save' method in WC_Product_Variable was almost identical
to the parent method in WC_Product, only adding a few bits
of extra processing before and after the actual datastore save.
This commit adds a couple of protected methods,
before_data_store_save_or_update and after_data_store_save_or_update,
and implements them in WC_Product_Variable so that the extra
'save' method in WC_Product_Variable can be removed.

Konamiman committed:
Remove superfluous 'delete' method in WC_Product_Variation class.

Konamiman committed:
Implement the product attributes lookup table data deletion.

Lookup entries for a product or a variation are deleted whenever
the product is deleted or trashed.

Konamiman committed:
Implement the product attributes lookup table data creation.

Lookup entries for a product or a variation are created whenever
the product is created, including untrashing and duplication
(not yet when an existing product is modified).

Konamiman committed:
Merge branch 'master' into add/29609

Konamiman committed:
Merge branch 'trunk' into add/29609

Konamiman committed:
Fix in FiltererTest after mrging from trunk.

Konamiman committed:
Implement handling of changesets in the LookupDataStore class.

Konamiman committed:
Disable attributes lookup table usage before initiating regeneration.

Konamiman committed:
Change LookupDataStore to allow updates being done in a scheduled action.

There's a new option, 'woocommerce_attribute_lookup__direct_updates'.
When set to 'yes', updates to the lookup table are performed as soon
as the change happen; otherwise, a scheduled action will do it,
the hook name is 'woocommerce_run_product_attribute_lookup_update_callback'
(the existing hook in the DataRegenerator class is renamed to
'woocommerce_run_product_attribute_lookup_update_callback')

Also, the settings page has a new "Advanced" section with a checkbox
to control the value of that new option; the section is visible only
when the feature has been enabled via LookupDataStore::show_feature.

Konamiman committed:
Add tests for product deletion in LookupDataStoreTest.

Konamiman committed:
Add tests for product changes changes in LookupDataStore.

Konamiman committed:
Add a lookup data regeneration for single product mechanism.

A new "Lookup data" metabox has been added to the product page
with a "Renegerate" button that regenerates the attributes lookup
data for that product.

Konamiman committed:
Move the UI to set/unset product attribute lookup table usage to settings.

The UI was previously in the tools page, it's now a setting under
products - advanced.

Konamiman committed:
Fix unit tests for the DataRegenerator class

Konamiman committed:
Fix unit tests for the LookupDataStore class

Konamiman committed:
Small adjustments to the per-product lookup data regeneration metabox.

roykho commented:
I tested all the scenarios and they seem to be working as described. I was not able to replicate this message though https://user-images.githubusercontent.com/937723/124109913-f0b7dc00-da67-11eb-93f9-f3947a98d044.png

I tried running the generation and quickly refresh the settings page to see but that message didn't show up. It could be it happened so fast. I don't know.

The other thing is for this section https://user-images.githubusercontent.com/937723/124111503-a7688c00-da69-11eb-923e-3019fecf0a6b.png

We're adding a significant UI to the products page but it seems this is more for developers that knows what they're doing. "Lookup Data" may not be understandable for store merchants. With that said, I don't think merchants will ever know to use that or worse they will ask support about it. May be we could move this functionality to elsewhere? What about to Status->Tools just like for the "Regenerate" lookup tables but have a text input field for them to add the product ID? I know it would be more work to get the product ID but at least we hide this functionality to a location that is more for dev debugging. Just a thought. Perhaps we could rope in other team members for opinions.

Konamiman commented:
Hi @roykho, about the regenerate button on the product page, I discussed this with @ObliviousHarmony and it seemed like a good idea as a mechanism to be used by store owners when suggested by HEs. But yeah, a second (or third or...) opinion would be great.

roykho commented on includes/abstracts/abstract-wc-product.php:
```diff
@@ -1374,13 +1376,22 @@ public function save() {
+		$this->after_data_store_save_or_update( $state );
+
+		// Update attributes lookup table if the product is new OR it's not but there are actually any changes.
+		if ( is_null( $changeset ) || ! empty( $changeset ) ) {
```
A little confused here. Aren't these two conditions contradicting? If it is null, it will never be not empty and vice versa?

Konamiman replied:
The `if` evaluates to false only if the changeset is an empty array and that's on purpose. I want the condition to be true when the product is new (so `null` changeset) or if there are changes (non-`null` and non-empty changeset), but I want it to be false when the product exists and there are no changes (so the changeset is an empty array).

Konamiman replied:
Do you think I could update the comment so that the intent is more clear?

roykho commented on includes/abstracts/abstract-wc-product.php:
```diff
@@ -1393,17 +1404,38 @@ public function save() {
+	 *
+	 * @return mixed A state value that will be passed to after_data_store_save_or_update.
+	 */
+	protected function before_data_store_save_or_update() {
```
I am assuming this is intentionally left blank for future changes?

Konamiman replied:
This is intentionally blank indeed, to be filled by derived classes if needed (as it happens in `class-wc-product-variable.php`).

roykho commented on includes/class-wc-product-variable.php:
```diff
@@ -449,47 +449,31 @@ public function validate_props() {
-		 */
-		do_action( 'woocommerce_before_' . $this->object_type . '_object_save', $this, $this->data_store );
-
+	protected function before_data_store_save_or_update() {
```
If we remove the `save` method, wouldn't that break backwards compatibility?

Konamiman replied:
The method exists already in the base class (`abstract-wc-product.php`). It can be removed here, thus removing some code suplication, thanks to the new `before_data_store_save_or_update` and `after_data_store_save_or_update` methods

roykho commented on includes/class-wc-product-variation.php:
```diff
@@ -583,21 +583,4 @@ protected function get_valid_tax_classes() {
-	 * @param  bool $force_delete Should the variation be deleted permanently.
-	 * @return bool result
-	 */
-	public function delete( $force_delete = false ) {
```
If we remove the `delete` method, wouldn't that break backwards compatibility?

Konamiman replied:
This method exists already in the base class (`abstract-wc-product.php`). I removed the method here because it's useless (it's just calling the method in the base class).

roykho reviewed (commented):
Here is my initial round of code review and left some comments.

ObliviousHarmony commented:
@roykho has a great point about this being a developer-focused utility.

Konamiman commented:
@ObliviousHarmony @roykho I still think that regenerating the lookup entries for a single product would be a tool not only helpful for developers but also for merchants, even if occasionally; and the product page seems the most natural place to put the UI to trigger it (adding a textbox for the product id elsewhere adds confusion and is more error-prone).

That said, maybe the UI I added can be improved. The metabox title could be something like "Advanced tools", it could be collapsed by default (if there's some way to do that), and the explanation of what the button does could be more detailed (maybe under a tooltip). Would that make sense?

roykho commented:
Hmm, technically if they click "Update" on the product page, doesn't that already trigger the regeneration? If that is the case, then this tool may be is not needed?

Konamiman commented:
@roykho Clicking "Update" will indeed execute `LookupDataStore::on_product_changed`, but this will schedule a regeneration or update of the lookup data for the product only if needed (full regeneration if product attributes have changed, only update if the stock status has changed, nothing at all for any other change). The tool is needed to force an unconditional regeneration.

roykho commented:
I see. Okay so I am generally ok with this PR, just let me know what we decide on the metabox and I will give it a final review before I merge it.

Konamiman committed:
Add missing JOIN parts from tax meta and meta queries.

The SQL query used to calculate the product counts using the product
attributes lookup table for the attribute filtering widget was missing
the JOIN parts coming from the tax meta and meta queries. The WHERE
parts however were being used, so the enitre query could fail if there
was a tax or meta query in place.

Konamiman committed:
Optimize the query to count products using the attributes lookup table.

The query to count products using the attributes lookup table for the
filter by attribute widget was adding an "AND term_ids in (...)"-type
subquery for each attribute participating in the filtering. Now at most
two such subqueries are generated, one for attributes configured for
OR type filtering and another one for the ones configured as AND; this
speeds up the query significantly when many attributes are used
simultaneously for the filtering.

Konamiman committed:
Remove the metabox for regenerating attribute lookup data for single product

Instead, a product selector has been added to the "Regenerate
product attributes lookup table" entry in the tools page. If a product
is selected, the tool regenerates the data only for that product;
otherwise, it regenerates the entire table.

This has forced a change on how the tools page is rendered. Now,
instead of each tool being just a description and a trigger link,
a form with GET method is rendered for each tool. The forms are rendered
first and then the tools, since HTML doesn't allow to include forms
inside tables; each button is associated to its form with a "form"
attribute.

Additionally, now the tools array returned by the woocommerce_debug_tools`
hook can have a 'selector' array with the details needed to render a
selector, which will also be part of the form for the tool.

Konamiman commented:
Hi @roykho (cc @ObliviousHarmony), I have replaced the metabox with a product selector in the "Regenerate the product attributes lookup table" tool. This has forced me to change the way the tool page is rendered, now one GET form is used for each tool; the details are in the last commit message.

roykho commented on src/Internal/ProductAttributesLookup/Filterer.php:
```diff
@@ -196,7 +196,9 @@ private function get_product_counts_query_using_lookup_table( $tax_query, $meta_
 		$query['select'] = 'SELECT COUNT(DISTINCT product_or_parent_id) as term_count, term_id as term_count_id';
 		$query['from']   = "FROM {$this->lookup_table_name}";
-		$query['join']   = "INNER JOIN {$wpdb->posts} ON {$wpdb->posts}.ID = {$this->lookup_table_name}.product_or_parent_id";
+		$query['join']   = "
```
With this format I believe the `INNER JOIN` won't get applied if `{$tax_query_sql['join']} {$meta_query_sql['join']}` is empty.

Konamiman replied:
You mean the line format? I put it that way for readability. If these are empty then only the `INNER JOIN {$wpdb->posts}` right below will apply.

roykho commented on includes/admin/views/html-admin-page-status-tools.php:
```diff
@@ -11,22 +11,46 @@
+foreach ( $tools as $action_name => $tool ) {
+	// phpcs:disable WordPress.Security.EscapeOutput.OutputNotEscaped
+	?>
+	<form id="<?php echo 'form_' . $action_name; ?>" method="GET" action="<?php echo esc_url( admin_url( 'admin.php?foo=bar' ) ); ?>">
```
Instead of ignoring the escape perhaps we should use `esc_attr()` here?

Konamiman replied:
I thought that was not necessary since all the attribute values are fixed and we know beforehand it's going to work and be secure. But sure, adding `esc_attr`s won't hurt.

roykho commented on includes/admin/views/html-admin-page-status-tools.php:
```diff
@@ -11,22 +11,46 @@
+		<?php wp_nonce_field( 'debug_action', '_wpnonce', false ); ?>
+		<input type="hidden" name="page" value="wc-status"/>
+		<input type="hidden" name="tab" value="tools"/>
+		<input type="hidden" name="action" value="<?php echo $action_name; ?>"/>
```
Same comment here as above about escaping.

roykho commented on src/Internal/ProductAttributesLookup/DataRegenerator.php:
```diff
@@ -345,11 +338,19 @@ private function add_initiate_regeneration_entry_to_tools_array( array $tools_ar
-		$this->initiate_regeneration();
+		if ( isset( $_REQUEST['regenerate_product_attribute_lookup_data_product_id'] ) ) {
+			$product_id = (int) $_REQUEST['regenerate_product_attribute_lookup_data_product_id'];
+			$this->check_can_do_lookup_table_regeneration( $product_id );
```
Small nitpick here but we probably can remove the `check` from the function name because `can` already means to check.

Konamiman replied:
I'd say that not exactly. A method name of `can_do_whatever` seems to imply that it's going to return a boolean, but in this case we are throwing an exception if the check fails; that's why I think the "check" in place is better, since it conveys the meaning off "I'll check if it's all ok" instead of "I'll tell you if it's all ok". That said, maybe "verify" or  "ensure" would be better than "check" in this case?

roykho commented on src/Internal/ProductAttributesLookup/DataRegenerator.php:
```diff
@@ -345,11 +338,19 @@ private function add_initiate_regeneration_entry_to_tools_array( array $tools_ar
+			$this->check_can_do_lookup_table_regeneration( $product_id );
+			$this->data_store->create_data_for_product( $product_id );
+		} else {
+			$this->check_can_do_lookup_table_regeneration();
```
Small nitpick here but we probably can remove the `check` from the function name because `can` already means to check.

roykho commented on src/Internal/ProductAttributesLookup/DataRegenerator.php:
```diff
@@ -367,64 +368,24 @@ private function enable_or_disable_lookup_table_usage( $enable ) {
-			! wc_get_product( $product_id )
-		) {
-			return;
+	private function check_can_do_lookup_table_regeneration( $product_id = null ) {
```
Small nitpick here but we probably can remove the `check` from the function name because `can` already means to check.

roykho commented on includes/admin/views/html-admin-page-status-tools.php:
```diff
@@ -11,22 +11,46 @@
+				</th>
+				<td class="run-tool">
+					<?php // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>
+					<input <?php echo ArrayUtil::is_truthy( $tool, 'disabled' ) ? 'disabled' : ''; ?> type="submit" form="<?php echo 'form_' . $action_name; ?>" class="button button-large" value="<?php echo esc_html( $tool['button'] ); ?>" />
```
`esc_html` should be `esc_attr` here.

roykho reviewed (changes requested):
Generally I am seeing the phpcs ignores on the escaping but not sure why? I left some comments on those and there are other areas as well but maybe if you can explain why they are not needed?

Konamiman committed:
Use 'esc_attr' instead of suppressing the PHPCS warning

roykho reviewed (approved):
Tested all scenarios in testing steps. LGTM!
